File: backend/app/routes/audit.py
from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.routing import Route

from ..database import get_db
from ..models import AuditLog, MeetingReservation
from ..schemas import AuditLogOut


async def list_audit_logs(request: Request):
    db = next(get_db())
    reservation_id = request.query_params.get("reservation_id")

    query = db.query(AuditLog)
    if reservation_id:
        query = query.filter(AuditLog.reservation_id == int(reservation_id))

    logs = query.order_by(AuditLog.created_at.desc()).all()
    result = [AuditLogOut.model_validate(log).model_dump(mode='json') for log in logs]
    return JSONResponse({"items": result})


async def get_failure_audits(request: Request):
    db = next(get_db())
    query = db.query(AuditLog).filter(
        AuditLog.action.in_(["return", "review_return", "reject"])
    ).order_by(AuditLog.created_at.desc())

    page = int(request.query_params.get("page", 1))
    page_size = int(request.query_params.get("page_size", 20))
    total = query.count()
    logs = query.offset((page - 1) * page_size).limit(page_size).all()

    result = []
    for log in logs:
        log_dict = AuditLogOut.model_validate(log).model_dump(mode='json')
        reservation = db.query(MeetingReservation).filter(
            MeetingReservation.id == log.reservation_id
        ).first()
        if reservation:
            log_dict["reservation_no"] = reservation.reservation_no
            log_dict["reservation_title"] = reservation.title
            log_dict["current_status"] = reservation.status
        result.append(log_dict)

    return JSONResponse({"total": total, "items": result, "page": page, "page_size": page_size})


routes = [
    Route("/", list_audit_logs, methods=["GET"]),
    Route("/failures", get_failure_audits, methods=["GET"]),
]

File: backend/src/routes/audit.js
import { Router } from 'express';
import { getDB } from '../models/database.js';

const router = Router();

router.get('/care-record/:careRecordId', (req, res) => {
  const db = getDB();
  const logs = db.prepare(`
    SELECT * FROM audit_logs WHERE care_record_id = ? ORDER BY created_at DESC
  `).all(req.params.careRecordId);
  res.json(logs);
});

router.get('/', (req, res) => {
  const db = getDB();
  const { action, actor_role, start_date, end_date, keyword } = req.query;

  let sql = 'SELECT * FROM audit_logs WHERE 1=1';
  const params = [];

  if (action) {
    sql += ' AND action = ?';
    params.push(action);
  }
  if (actor_role) {
    sql += ' AND actor_role = ?';
    params.push(actor_role);
  }
  if (start_date) {
    sql += ' AND created_at >= ?';
    params.push(start_date);
  }
  if (end_date) {
    sql += ' AND created_at <= ?';
    params.push(end_date);
  }
  if (keyword) {
    sql += ' AND (detail LIKE ? OR reason LIKE ? OR actor_name LIKE ?)';
    params.push(`%${keyword}%`, `%${keyword}%`, `%${keyword}%`);
  }

  sql += ' ORDER BY created_at DESC LIMIT 200';

  const logs = db.prepare(sql).all(...params);
  res.json(logs);
});

export default router;
